### ISSUE TITLE
websocket-send: Cannot send message to a closed websocket

### ISSUE THREAD

**Issue by chanGimeno** [millejoh/emacs-ipython-notebook#22]
I am using IPython 2.3.1 and EIN form MELPA. I can open the notebook list `M-x ein:notebooklist-open RET RET`. I can also open a notebook, but then I cannot execute any cell.

## Check list

- [x] Read the "Avoiding Common Emacs Traps" section in
  https://github.com/millejoh/emacs-ipython-notebook/blob/master/CONTRIBUTING.md
- [x] Does IPython works from the web browser?
- [x] Confirm the problem was not due to badly compiled
  files by removing all the `*.elc` files from source directory of EIN and
  its dependencies.
- [x] Confirm that the libraries are loaded as expected and are at the required version.
  (You can check the location in the "System info" section below)

## Description of the problem you have
I am using IPython 2.3.1 and EIN form MELPA. I can open the notebook list `M-x ein:notebooklist-open RET RET`. I can also open a notebook, but then I cannot execute any cell.

## Steps to reproduce the problem

1. In a shell `ipython notebook --ip=127.0.0.1`
2. `M-x ein:notebooklist-open RET RET`
3. Open a notebook
4. M-RET in a notebook cell

## Expected output
Executed python code in the notebook cell

## Your EIN configuration (in .emacs.d/init.el or somewhere else)
```
;; ELPA packages
(add-to-list 'load-path "~/.emacs.d/")
```

## Your IPython configuration

1. What is your IPython version? (run `ipython --version`):
   2.3.1

2. How do you start2.3. IPython? (e.g., `ipython notebook --port 9999`):
   `ipython notebook --ip=127.0.0.1`

3. What is your IPython notebook port number or URL?:
   8888

###### *Warnings*
```
Error (websocket): in callback `on-open': Cannot send message to a closed websocket: cl-struct-websocket-frame, text, "{\"header\":{\"msg_id\":\"316f999e-2fdc-4529-80e5-24c8bc6b0c74\", \"username\":\"username\", \"session\":\"ba9991c8-3e8e-4103-aded-474047269bef\", \"msg_type\":\"execute_request\"}, \"metadata\":{}, \"content\":{\"code\":\"__import__('sys').stdout.write(__import__('os').getcwd())\", \"silent\":true, \"user_variables\":[], \"user_expressions\":{}, \"allow_stdin\":false}, \"parent_header\":{}}", nil, t
Error (websocket): in callback `on-open': Wrong type argument: arrayp, nil
Error (websocket): in callback `on-close': Wrong type argument: arrayp, nil
```

###### *Messages*
```
websocket-send: Cannot send message to a closed websocket
Blocking call to accept-process-output with quit inhibited!! [2 times]
REQUEST [error] Error (error) while connecting to http://127.0.0.1:8888/api.
ein: [info] Opened notebook list at 8888 with path 
```

###### SHELL
```
2015-01-30 14:51:37.043 [NotebookApp] Using existing profile dir: u'/users/gime_se/.ipython/profile_default'
2015-01-30 14:51:37.058 [NotebookApp] Using MathJax from CDN: https://cdn.mathjax.org/mathjax/latest/MathJax.js
2015-01-30 14:51:37.070 [NotebookApp] Serving notebooks from local directory: /users/gime_se
2015-01-30 14:51:37.071 [NotebookApp] 0 active kernels 
2015-01-30 14:51:37.071 [NotebookApp] The IPython Notebook is running at: http://127.0.0.1:8888/
2015-01-30 14:51:37.071 [NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).

(process:5487): GLib-CRITICAL **: g_slice_set_config: assertion 'sys_page_size == 0' failed
2015-01-30 14:53:05.855 [tornado.access] WARNING | 404 GET /api (127.0.0.1) 4.81ms referer=None
2015-01-30 14:53:05.982 [tornado.access] 302 GET /api/notebooks/ (127.0.0.1) 0.39ms
2015-01-30 14:53:07.861 [NotebookApp] WARNING | Notebook /session_band_4_o2a.ipynb is not trusted
2015-01-30 14:53:08.035 [NotebookApp] Kernel started: c5a23594-9ee4-4bb8-9d3e-ca15aee087fb
2015-01-30 14:53:08.076 [NotebookApp] WARNING | Missing Origin header, rejecting WebSocket connection.
2015-01-30 14:53:08.076 [tornado.application] ERROR | Uncaught exception in /api/kernels/c5a23594-9ee4-4bb8-9d3e-ca15aee087fb/shell
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/tornado/websocket.py", line 303, in wrapper
    return callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/IPython/html/base/zmqhandlers.py", line 147, in open
    raise web.HTTPError(403)
HTTPError: HTTP 403: Forbidden
```
###### Software versions
Python 2.7.6
tornado 3.1
pyzmq 13.0.0

## System info:

```cl
("EIN system info"
 :emacs-version
 "GNU Emacs 24.3.1 (x86_64-suse-linux-gnu, GTK+ Version 3.10.9)
 of 2014-11-12 on cloud104"
 :emacs-bzr-version nil
 :window-system x
 :emacs-variant nil
 :os
 (:uname
  "Linux arete 3.11.10-25-default #1 SMP Wed Dec 17 17:57:03 UTC 2014 (8210f77) x86_64 x86_64 x86_64 GNU/Linux
"
  :lsb-release
  "LSB Version:	n/a
Distributor ID:	openSUSE project
Description:	openSUSE 13.1 (Bottle) (x86_64)
Release:	13.1
Codename:	Bottle
")
 :image-types
 (svg imagemagick png gif tiff jpeg xpm xbm pbm postscript)
 :image-types-available
 (svg imagemagick png gif tiff jpeg xpm xbm pbm postscript)
 :request
 (:backend curl)
 :ein
 (:version "0.3"
	   :source-dir "/users/gime_se/.emacs.d/elpa/ein-20150116.2001/")
 :lib
 ((:name "websocket"
	 :path "~/.emacs.d/elpa/websocket-20141226.1957/websocket.elc"
	 :featurep t
	 :version-var websocket-version
	 :version "1.3")
  (:name "request"
	 :path "~/.emacs.d/elpa/request-20140316.417/request.elc"
	 :featurep t
	 :version-var request-version
	 :version "0.2.0")
  (:name "auto-complete"
	 :path "~/.emacs.d/elpa/auto-complete-20150124.1523/auto-complete.elc"
	 :featurep nil
	 :version-var nil
	 :version nil)
  (:name "auto-complete"
	 :path "~/.emacs.d/elpa/auto-complete-20150124.1523/auto-complete.elc"
	 :featurep nil
	 :version-var nil
	 :version nil)
  (:name "popup"
	 :path "~/.emacs.d/elpa/popup-20150116.1223/popup.elc"
	 :featurep nil
	 :version-var nil
	 :version nil)
  (:name "fuzzy"
	 :path "~/.emacs.d/elpa/fuzzy-20131025.2343/fuzzy.elc"
	 :featurep nil
	 :version-var nil
	 :version nil)
  (:name "pos-tip"
	 :path "~/.emacs.d/elpa/pos-tip-20141122.512/pos-tip.elc"
	 :featurep nil
	 :version-var nil
	 :version nil)
  (:name "python"
	 :path "/usr/share/emacs/24.3/lisp/progmodes/python.elc"
	 :featurep t
	 :version-var nil
	 :version nil)
  (:name "python-mode"
	 :path "~/.emacs.d/elpa/python-mode-20150117.955/python-mode.el"
	 :featurep nil
	 :version-var nil
	 :version nil)
  (:name "markdown-mode"
	 :path "~/.emacs.d/elpa/markdown-mode-20150121.1229/markdown-mode.elc"
	 :featurep nil
	 :version-var nil
	 :version nil)
  (:name "smartrep"
	 :path "~/.emacs.d/elpa/smartrep-20140404.1714/smartrep.elc"
	 :featurep nil
	 :version-var nil
	 :version nil)))
```

**Comment by cgcgcg** [comment]
If I remember correctly, I managed to fix that issue by updating tornado..

**Comment by millejoh** [comment]
Yep - almost there! You need to Tornado 4.0.2 or higher. The debugging from should do a better job communicating this simple, but easily overlooked fix.

The documentation has the full list of [requirements](http://millejoh.github.io/emacs-ipython-notebook/#requirements).

**Comment by chanGimeno** [comment]
Thanks for the tips and sorry for overseeing the Tornado version! Now, I have a working EIN :-)

Actually, I indeed tried the upgraded tornado version before, but I couldn't even run `ipython notebook` without error messages, so I decided to come back to a working `IPython` environment. I believe this problem is related to openSUSE 13.1, which only provides IPython 1.0.0 (and also older versions of tornado, zmq, ...). 

For me to have a working IPython 2.x together with EIN in openSUSE 13.1 I had to do the following steps:

1. add devel:languages:python repository
```
    zypper addrepo http://download.opensuse.org/repositories/devel:languages:python/openSUSE_13.1/devel:languages:python.repo
    zypper refresh
```
2.  `zypper install python-tornado-4.0.2-6.1.x86_64`
    which also installed `python-backports.ssl_match_hostname python-futures python-pycares`
3.  `zypper rm python-pyzmq`
4.  `zypper in python-pyzmq-14.3.1-45.3`
    (which also installed `libzmq4 python-pyzmq`)

After that, I could run the notebook list, the worksheets and most of the stuff.

The only couple of minor things that I have noticed are:

* Open a regular IPython console `M-x ein:console-open [C-c C-o]` doesn't work:
```
[ZMQTerminalIPythonApp] CRITICAL | Could not find existing kernel connection file kernel-094f2a3e-fb75-407c-b20f-da77efeeab01.json

Process Python[*ein: 8888/session_band_4_o2a.ipynb*] exited abnormally with code 1
```
* `[C-c C-u] ein:worksheet-change-cell-type` doesn't work:
```
Stack overflow in equal
```
* Interestingly, `[C-c C-t] ein:worksheet-change-cell-type` does work!

Thanks for the support and development effort! EIN is really an amazing tool!!!
